lib/Sema/ConstraintSystem.cpp

```diff
@@ -1391,6 +1391,29 @@ void ConstraintSystem::addOverloadSet(Type boundType,
     return;
   }
 
+  // Performance hack: if there are two generic overloads, and one is
+  // more specialized than the other, prefer the more-specialized one.
+  if (!favoredChoice && choices.size() == 2 &&
```

Comment:
The idea is great! I'm trying to implement similar thing as well but instead of hard-coding this to 2 overloads only, maybe it would be better to try and group overload choices based on declaration context and rank them in groups so at the end there are N favored choices in the set?... Another idea might be to move from favoring to making it a ranking criteria since we erase worse intermediate solutions...

Reply:
You're absolutely right that we could generalize this. In general, if an one overload is more specialized than another, we should try it first. We could probably encode this information in an tree of disjunctions using favored constraints... but it would probably be better in the long run to compute the lattice describing the relationships and have the solver take an appropriate path through the lattice.

As for moving it to ranking criteria, that won't prune as efficiently, because we often still explore those other options. I'd rather